keps/sig-cluster-lifecycle/kubeadm/20190916-kubeadm-operator.md

```diff
@@ -0,0 +1,363 @@
+  selector:
+    matchLabels:
+      app: a
+  template:
+    metadata:
+      labels:
+        app: a
+    alphaCertsRenew: {}
+```
+
+It is important to notice that the  `Action controller` is going to always generate `TaskGroup` in sequential order - only after the current `TaskGroup` is completed successfully a new one will be generated.
+
+> during implementation of the first proof of concept we might consider to extend `nodeSelector` semantic in order to express criteria like e.g. the first control plane node
+
+> during implementation of the first proof of concept we might consider to extend `Task` template semantic in order to allow to execute more than one command.
+
+> during implementation of the first proof of concept we might consider to implement a lock mechanisms for prevventing concurrent execution of more than one `Action` at any time.
+
+> Alternative name for the `TaskGroup` might be `TaskBatch` (batch = a quantity or consignment of goods produced at one time).
+> Instead usage of `TaskSet` or `TaskDeployment` was discarded because of its generally imply something that creates Pods, which it is not the case.
+
+
+#### TaskGroupController and Task
+The `TaskGroup controller` will be responsible for managing the  `TaskGroup` CRD and implements all the logic for generating `Task` from a `TaskGroup` according to the current topology of the Kubernetes cluster.
+
+Each generated `Task` is expected to target a specific `Node` in the cluster.
+
+```yaml
+apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
+kind: Task
+metadata:
+  name: task-sample-control-plane-1
+  labels:
+    app: a
+spec:
+  nodeName: control-plane-1
+  alphaCertsRenew: {}
+```
+
+or
+
+```yaml
+apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
+kind: Task
+metadata:
+  name: task-sample-control-plane-2
+  labels:
+    app: a
+spec:
+  nodeName: control-plane-2
+  alphaCertsRenew: {}
+```
+
+The `TaskGroup controller` is going to generate `Task` according to a policy that by default is sequential order - only after the current `Task` is completed successful a new one will be generated.
+
+In future releases of the kubeadm operator also additional policy could be added for supporting e.g. parallel execution of `Task` or disruption budget (max N `Task` in executed at the same time).
+
+In any case, the `TaskGroup controller` is going to ensure a predictable execution order based on the alphabetical order of `Node`s in the scope of a `TaskGroup`.
+
+#### TaskController
+
+The `Task controller` is the last component of the kubeadm operator. 
+
+The main characteristic of this controller is the fact that this controller is deployed on all the nodes in the cluster using a `DaemonSet`, and that each instance is responsible for managing `Task` targeting a specific node.
+
+`Task` reconciliation triggers the execution of kubeadm commands on such node; in order to do so, `Task controller` require to be executed in privileged mode.
+
+#### Observability and Operability
```

Comment:
I think we should add a dedicated subsection for observability, for log management event. It shouldn't be detailed, but just discuss how would a user troubleshoot a failed action/taskGroup

Reply:
Good suggestion, done!